### PR TITLE
[thread-cert] refactor case 5.8.2 and 5.8.3 using pktverify

### PR DESCRIPTION
**Pull request by FiveDimensions**
based on #4428

### CODE REVIEW

**Comment by simonlingoogle** [comment]
Thanks, @FiveDimensions ! 
Great to see more test cases being packet verified.

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 1]
```diff
@@ -85,56 +85,55 @@ def verify(self, pv):
 
         # Step 1: The DUT must start the network using
         # thrKeySequenceCounter = 0
-        _lpkts = leader_pkts.filter_mle_cmd(consts.MLE_ADVERTISEMENT).must_next()
+        _lpkts = leader_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:88]: `must_next()` returns a single Packet
```suggestion
        lpkt = leader_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
```

BTW, what does `lpkts` stands for?

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 2]
```diff
@@ -85,56 +85,55 @@ def verify(self, pv):
...
-        _rpkts.filter_mle_cmd(consts.MLE_ADVERTISEMENT).must_next()
-        [_,i] = _rpkts.index
-        self.assertTrue(_rpkts[i-1].wpan.aux_sec.key_index == 1 and
-                        _rpkts[i-1].wpan.aux_sec.key_id_mode == 2 and
-                        _rpkts[i-1].wpan.aux_sec.key_source == 0)
+        _rpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        [_, i] = _rpkts.index
+        self.assertTrue(_rpkts[i - 1].wpan.aux_sec.key_index == 1 and
+                        _rpkts[i - 1].wpan.aux_sec.key_id_mode == 2 and
+                        _rpkts[i - 1].wpan.aux_sec.key_source == 0)
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:102]:
```suggestion
        _rpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next().must_verify("""
            wpan.aux_sec.key_index == 1 and
            wpan.aux_sec.key_id_mode == 2 and 
            wpan.aux_sec.key_source == 0
        """)
```

Usually, we don't need to get the `index` manually. 

`assertTrue` works, but better call `must_verify`.

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 3]
```diff
@@ -85,56 +85,55 @@ def verify(self, pv):
...
 
         # Step 3: Leader send an ICMPv6 Echo Request to Router_1.
         # The MAC Auxiliary security header must contain
         # KeyIndex = 1, KeyID Mode = 1
         _lpkts.filter_ping_request().must_next()
-        [_,i] = _lpkts.index
-        self.assertTrue(_lpkts[i-1].wpan.aux_sec.key_index == 1 and
-                        _lpkts[i-1].wpan.aux_sec.key_id_mode == 1)
+        [_, i] = _lpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:108]: ditto. 

Use pattern: `pkts.filter_xxx().must_next().must_verify(...)`

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 4]
```diff
@@ -85,56 +85,55 @@ def verify(self, pv):
...
+                        _lpkts[i - 1].wpan.aux_sec.key_id_mode == 1)
 
         # Step 4: Router_1 send an ICMPv6 Echo Reply to Leader.
         # The MAC Auxiliary security header must contain
         # KeyIndex = 1, KeyID Mode = 1
         _rpkts.filter_ping_reply().must_next()
-        [_,i] = _rpkts.index
-        self.assertTrue(_rpkts[i-1].wpan.aux_sec.key_index == 1 and
-                        _rpkts[i-1].wpan.aux_sec.key_id_mode == 1)
+        [_, i] = _rpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:116]: ditto

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 5]
```diff
@@ -85,56 +85,55 @@ def verify(self, pv):
...
         # Step 5: The DUT MUST set incoming frame counters to 0 for all existing devices.
 
         # Step 6: Leader Send an ICMPv6 Echo Request to Router_1.
         # The MAC Auxiliary security header must contain
         # KeyIndex = 2, KeyID Mode = 1
         _lpkts.filter_ping_request().must_next()
-        [_,i] = _lpkts.index
-        self.assertTrue(_lpkts[i-1].wpan.aux_sec.key_index == 2 and
-                        _lpkts[i-1].wpan.aux_sec.key_id_mode == 1)
+        [_, i] = _lpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:126]: ditto

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 6]
```diff
@@ -85,56 +85,55 @@ def verify(self, pv):
...
+                        _lpkts[i - 1].wpan.aux_sec.key_id_mode == 1)
 
         # Step 7: Router_1 send an ICMPv6 Echo Reply to Leader.
         # The MAC Auxiliary security header must contain
         # KeyIndex = 2, KeyID Mode = 1
         _rpkts.filter_ping_reply().must_next()
-        [_,i] = _rpkts.index
-        self.assertTrue(_rpkts[i-1].wpan.aux_sec.key_index == 2 and
-                        _rpkts[i-1].wpan.aux_sec.key_id_mode == 1)
+        [_, i] = _rpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:134]: ditto

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 7]
```diff
@@ -86,38 +86,37 @@ def verify(self, pv):
...
         _lpkts = leader_pkts.range(leader_pkts.index)
         _rpkts = router_pkts.range(leader_pkts.index)
 
-        _rpkts.filter_mle_cmd(consts.MLE_ADVERTISEMENT).must_next()
-        [_,i] = _rpkts.index
-        self.assertTrue(_rpkts[i-1].wpan.aux_sec.key_index == 128 and
-                        _rpkts[i-1].wpan.aux_sec.key_id_mode == 2 and
-                        _rpkts[i-1].wpan.aux_sec.key_source == 127)
+        _rpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        [_, i] = _rpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:100]: ditto

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 8]
```diff
@@ -86,38 +86,37 @@ def verify(self, pv):
 
         # Step 1: The DUT must start the network using
         # thrKeySequenceCounter = 127
-        _lpkts = leader_pkts.filter_mle_cmd(consts.MLE_ADVERTISEMENT).must_next()
+        _lpkts = leader_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:89]:
```suggestion
        pkt = leader_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
```
`must_next` returns a single Packet.

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 9]
```diff
@@ -62,16 +66,76 @@ def test(self):
...
+        _rpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        [_, i] = _rpkts.index
+        self.assertTrue(_rpkts[i - 1].wpan.aux_sec.key_index == 128 and
+                        _rpkts[i - 1].wpan.aux_sec.key_id_mode == 2 and
+                        _rpkts[i - 1].wpan.aux_sec.key_source == 127)
+
+        # Step 3: Leader send an ICMPv6 Echo Request to Router_1.
+        # The MAC Auxiliary security header must contain
+        # KeyIndex = 128, KeyID Mode = 1
+        _lpkts.filter_ping_request().must_next()
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:108]: ditto

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 10]
```diff
@@ -62,16 +66,76 @@ def test(self):
...
+        [_, i] = _lpkts.index
+        self.assertTrue(_lpkts[i - 1].wpan.aux_sec.key_index == 128 and
+                        _lpkts[i - 1].wpan.aux_sec.key_id_mode == 1)
+
+        # Step 4: Router_1 send an ICMPv6 Echo Reply to Leader.
+        # The MAC Auxiliary security header must contain
+        # KeyIndex = 128, KeyID Mode = 1
+        _rpkts.filter_ping_reply().must_next()
+        [_, i] = _rpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:117]: ditto

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 11]
```diff
@@ -62,16 +66,76 @@ def test(self):
...
+                        _rpkts[i - 1].wpan.aux_sec.key_id_mode == 1)
+
+        # Step 5: The DUT MUST implementation specific means increment
+        # thrKeySequenceCounter by 1 to force a key switch
+
+        # Step 6: Leader Send an ICMPv6 Echo Request to Router_1.
+        # The MAC Auxiliary security header must contain
+        # KeyIndex = 1, KeyID Mode = 1
+        _lpkts.filter_ping_request().must_next()
+        [_, i] = _lpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:128]: ditto

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 12]
```diff
@@ -62,16 +66,76 @@ def test(self):
...
+        self.assertTrue(_lpkts[i - 1].wpan.aux_sec.key_index == 1 and
+                        _lpkts[i - 1].wpan.aux_sec.key_id_mode == 1)
+
+        # Step 7: Router_1 send an ICMPv6 Echo Reply to Leader.
+        # The MAC Auxiliary security header must contain
+        # KeyIndex = 1, KeyID Mode = 1
+        _rpkts.filter_ping_reply().must_next()
+        [_, i] = _rpkts.index
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:136]: ditto

File: tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py [hunk 13]
```diff
@@ -62,16 +66,76 @@ def test(self):
...
+
+        LEADER = pv.vars['LEADER']
+        ROUTER = pv.vars['ROUTER']
+        leader_pkts = pkts.filter_wpan_src64(LEADER)
+        router_pkts = pkts.filter_wpan_src64(ROUTER)
+
+        # Step 1: The DUT must start the network using
+        # thrKeySequenceCounter = 127
+        _lpkts = leader_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        self.assertTrue(_lpkts.wpan.aux_sec.key_source == 127)
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:90]: use `must_verify`

Comment (FiveDimensions) [reply tests/scripts/thread-cert/Cert_5_8_03_KeyIncrementRollOver.py:95]: done

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 14]
```diff
@@ -61,16 +65,75 @@ def test(self):
...
+        # Step 1: The DUT must start the network using
+        # thrKeySequenceCounter = 0
+        _lpkts = leader_pkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        self.assertTrue(_lpkts.wpan.aux_sec.key_source == 0)
+
+        # Step 2: Verify that the topology described above is created.
+        # MLE Auxiliary security header shall contain Key Source = 0,
+        # KeyIndex = 1, KeyID Mode = 2
+        leader_pkts.filter_mle_cmd(MLE_CHILD_ID_RESPONSE).must_next()
+        _lpkts = leader_pkts.range(leader_pkts.index)
```
[the remark this answers is not in the document]
Comment (FiveDimensions) [reply tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:100]: @simonlingoogle  I need to start from the leader_pkts.index and create a new packetfilter _lpkts, but copy() only duplicate a whole leader_pkts,  thoughts?

[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py:100]: copy duplicate the filter with the current index. So, should be more or less same as `range(leader_pkts.index)`. The only difference is that `copy()` set `cascade=False` by default.

File: tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py [hunk 15]
```diff
@@ -61,16 +65,75 @@ def test(self):
...
+        # KeyIndex = 1, KeyID Mode = 1
+        _lpkts.filter_ping_request().must_next()
+        [_, i] = _lpkts.index
+        self.assertTrue(_lpkts[i - 1].wpan.aux_sec.key_index == 1 and
+                        _lpkts[i - 1].wpan.aux_sec.key_id_mode == 1)
+
+        # Step 4: Router_1 send an ICMPv6 Echo Reply to Leader.
+        # The MAC Auxiliary security header must contain
+        # KeyIndex = 1, KeyID Mode = 1
+        _rpkts.filter_ping_reply().must_next()
```
[the remark this answers is not in the document]
Comment (FiveDimensions) [reply tests/scripts/thread-cert/Cert_5_8_02_KeyIncrement.py]: used leader mleid to filter : _rpkts.filter(lambda p: p.ipv6.dst == leader_mleid).filter_ping_reply( ).must_next().must_verify(lambda p: p.wpan.aux_sec.key_index == 1 and p. wpan.aux_sec.key_id_mode == 1)

**Pull request merged**